{% extends '_layouts/base.html' %}

{% load crispy_forms_tags %}
{% load django_bootstrap_breadcrumbs %}

{% block css %}
	{% load static from staticfiles %}
		<!-- pre-define in-house css for main page-->
		<link rel="stylesheet" href="{% static 'css/main-page.css' %}" type="text/css" />
{% endblock css %}

<!-- Breadcrumbs-->
{% block breadcrumbs %}
	{{ block.super }}
	{% breadcrumb "References" "references" %}
	{% breadcrumb "Add reference" "reference-add" %}
{% endblock %}

<!-- Main content-->
{% block content %}
	<p class="insa-title"><a class="btn btn-primary" data-toggle="collapse" href="#collapseExample" aria-expanded="false" aria-controls="collapseExample">Guide to generate additional references</a></p>
	<div class="collapse bg-insa-fixed-top" id="collapseExample">
  		<p class="insa-description-read-more-first">If you are using the suggested wet-lab pre-NGS protocol (link) and you want to compare your sequences against a reference available 
  		at INSaFLU database, <strong>you do not need to upload a reference sequence</strong>.</p>
  		<p class="insa-description-read-more">However, if you want to upload additional reference files <strong>(“.fasta” extension; maximum 20000 bp per file)</strong>
  		to the user-restricted reference database, you must have:</p>
  		<ul>
	  		<li><p class="mb-0 insa-description-read-more">multi-FASTA files containing the set of reference sequences that constitute the influenza “whole-genome” sequence 
	  		of a particular virus (e.g, the combination of the traditional 8 amplicons targeting the 8 eight influenza RNA segments).</p></li>
	  		<li><p class="mb-0 insa-description-read-more">Each individual sequence must have the precise size of each “intra-amplicon” target sequence that 
	  		you capture by each one of the RT-PCR amplicons.</p></li>
	  		<li><p class="mb-0 insa-description-read-more">INSaFLU automatically annotates uploaded multi-FASTA sequences upon submission, but, if you prefer, 
	  		you can also upload (optionally) the respective multi-GenBank file.</p></li>
	  	</ul>
  		<p class="insa-description-read-more-header"><strong>Important notes to generate additional reference sequences:</strong></p>
  		<ul>
	  		<li><p class="insa-description-read-more">(multi) FASTA format is widely applied to save either nucleotide sequences or peptide sequences. An easy way to 
	  		handle/generate multi-FASTA files is by opening a text file (e.g., NOTEPAD) and paste individual sequences after each header line. The FASTA IDs (after 
	  		the '>' character) represent the individual sequence names. For the sake of simplicity, you may designate each sequence as 1, 2, 3, 4, 5, 6 , 7 and 8, 
	  		following the traditional influenza segments order (keeping this numerical order is advisable). At the end, you just have to save the multi-FASTA file 
	  		as “.fasta” (please avoid symbols or blank spaces in the file names).</p></li>
	  		<li><p class="insa-description-read-more">you may generate your multi-FASTA files in order to fit your amplicon schema by simply adjusting (i.e., 
	  		the whole-genome sequences available for download at INSaFLU or at influenza-specific sequence repositories, such as the 
	  		<a href="https://www.fludb.org">Influenza Research Database</a>, <a href="https://www.ncbi.nlm.nih.gov/genomes/FLU/Database/nph-select.cgi?go=database">NCBI Influenza Virus Resource</a> 
	  		and <a href="https://www.gisaid.org/">EpiFLU/GISAID</a>.</p></li>
	  		<li><p class="insa-description-read-more-last">INSaFLU requires reference sequences exclusively composed by <strong>non-degenerate bases (i.e. A, T, C, or G)</strong>. 
	  		As such, please <strong>ensure that all degenerated bases (e.g., R, Y, M, K, S and W) are replaced</strong> by non-degenerate sequences before uploading. <mark>The choice 
	  		of the base used in the replacement (e.g., “A” or “G” when replacing an “R”) has no impact on the analysis.</mark> It simply means that mutations falling 
	  		in the replaced nucleotide position will be reported taking into account the reference base selected.</p></li>
	  	</ul>
  	</div>
	{% crispy form %}
{% endblock %}
