{% extends '_layouts/base.html' %}

{% load static from staticfiles %}

{% block css %}
	{% load static from staticfiles %}
		<!-- pre-define in-house css for main page-->
		<link rel="stylesheet" href="{% static 'css/main-page.css' %}" type="text/css" />
{% endblock css %}

{% block headline %}
{% endblock headline %}

{% block content %}
	<div class="container-fluid">
		<div class="row">
			<div class="col-sm-8">
				<div class="insa-header">
			        	<h1 class="insa-title" style="margin:10px">From “reads” to WGS-based first-line influenza laboratory surveillance</h1>
			        	<p class="insa-description bg-insa-fixed-top" style="">INSaFLU (“INSide the FLU”) is an influenza-specific bioinformatics free web-based suite that deals
								  with primary data (reads) towards the automatic generation of the output data that are actually the core first-line 
								  “genetic requests” for effective and timely influenza laboratory surveillance (e.g., type and sub-type, gene and 
								  whole-genome consensus sequences, variants annotation, alignments and phylogenetic trees). Data integration is continuously 
								  scalable, fitting the need for a real-time epidemiological surveillance during the flu epidemics.</p>
						<p class="insa-title"><a class="btn btn-primary" data-toggle="collapse" href="#collapseExample" aria-expanded="false" aria-controls="collapseExample">Read More</a></p>
						<div class="collapse bg-insa-fixed-top" id="collapseExample">
  							<p class="insa-description-read-more-first">The INSaFLU bioinformatics pipeline currently consists of 6 core steps (see Figure). For more details about INSaFLU, please read <strong>Borges V,
  							Pinheiro M et al. 2018</strong>, <a href="#">bioRxiv</a>. A detailed tutorial of INSaFLU usage, as well as documentation about the current software settings and versions,
  							is provided <a href="#">here</a>.</p>
  							
  							<img src="{% static 'insa/CORE_STEPS.png' %}" style="width:450px; margin-left:auto; margin-right:auto; display: block;" id='insa_logo_id'>
  							<p class="insa-description-read-more-header"><strong>Read quality analysis and improvement</strong></p>
  							<p class="insa-description-read-more">This first step automatically analyzes and improves the quality of the uploaded raw sequencing data (reads). It generates improved NGS 
  							data and graphical quality control reports.</p>
  							
  							<p class="insa-description-read-more-header"><strong>Type and sub-type detection</strong></p>
							<p class="insa-description-read-more">INSaFLU automatically detects the influenza type and sub-type/lineage of each sample upon data submission. This typing data guides 
							the subsequent downstream module and constitutes an optimal complement to the traditional real-time RT-PCR assays, as INSaFLU is able to discriminate all currently defined
							18 hemagglutinin subtypes, 11 neuraminidase sub-types and influenza B lineages.</p>
							<p class="insa-description-read-more-header"><strong>Variant detection and consensus generation</strong></p>
							<p class="insa-description-read-more">This module provides the first-line "genetic data" for seasonal influenza laboratory surveillance, i.e. the list of variants
							(SNPs and indels) and their effect at protein level and also consensus sequences (at both the locus and whole-genome levels). The latter constitutes the whole basis
							for the downstream phylogenetic inferences driving the continuous tracking of influenza temporal/geographical spread.</p>
							<p class="insa-description-read-more-header"><strong>Coverage analysis</strong></p>
							<p class="insa-description-read-more">INSaFLU automatically provides a deep analysis of the vertical and horizontal coverage per sample (for each amplicon),
							by generating several coverage statistics and also graphical plots of the fluctuation of the depth of coverage throughout each amplicon region. This step enables
							the inspection of the technical success and the potential unveiling of relevant genetic events, such as reassortment or homologous recombination.</p>
							<p class="insa-description-read-more-header"><strong>Alignment/phylogeny</strong></p>
							<p class="insa-description-read-more">INSaFLU dynamically builds ready-to-explore nucleotide/amino acid sequence alignments and phylogenetic trees at both locus and
							whole-genome scale. These outputs are automatically re-build and updated as more samples are added to user-restricted INSaFLU projects, making continuous data integration
							completely flexible and scalable. This module can also be run independently over a set of user-selected sequences so that phylogenetic diversity of circulating viruses
							can be better evaluated and integrated according with users’ specific needs. Alignments and trees can be explored in situ or through multiple compatible downstream
							applications for fine-tune data analysis.</p>
							<p class="insa-description-read-more-header"><strong>Intra-host minor variant detection (and uncovering of putative mixed infections)</strong></p>
							<p class="insa-description-read-more-last">Influenza evolution has historically been inferred from consensus sequences representing the dominant virus lineage within
							each infected host at a particular instant, which has limited our knowledge on intra-patient virus population diversity and transmission dynamics. In this context,
							INSaFLU additionally provides the user the possibility to get insight on the influenza intra-patient sub-population dynamics by the scrutiny and annotation of minor
							intra-host single nucleotide variants (iSNVs). INSaFLU also automatically flags samples as “putative mixed infections” if the population admixture enrolls influenza
							viruses presenting clearly distinct genetic backgrounds.</p> 
						</div>
			    </div>
			    
			    <div class="card">
					<div class="card-block">
						<h4 class="card-title">Future Directions</h4>
						<p class="card-text">INSaFLU is under active development in order to have additional features, such as modules to automatically 
						detect virus reassortment and to perform temporal and geographical data integration and visualization.</p>
					<!-- 	<a href="#" class="card-link">Card link</a>
						<a href="#" class="card-link">Another link</a> -->
					</div>
				</div>
				<p> </p>
				<div class="card">
					<div class="card-block">
						<h4 class="card-title">How to cite</h4>
						<p class="card-text">If you use INSaFLU in your work, please cite this publication:</p>
					 	<a href="#" class="card-link">Borges V, Pinheiro M et al. 2018, bioRxiv</a>
					</div>
				</div>
				<p> </p>
				<div class="card">
					<div class="card-block">
						<h4 class="card-title">Contacts</h4>
						<p class="card-text">If you have any questions, comments or suggestions, please contact us:</p>
					 	<a href="mailto:vitor.borges@insa.min-saude.pt?subject=InsaFlu" class="card-link">vitor.borges at insa.min-saude.pt</a>
					 	<a href="mailto:j.paulo.gomes@insa.min-saude.pt?subject=InsaFlu" class="card-link">j.paulo.gomes at insa.min-saude.pt</a>
					</div>
				</div>
				
				<p> </p>
				<div class="row">
					<img src="{% static 'insa/Republica_Portuguesa_logo.svg' %}" style="width:180px; padding-right:30px; padding-top:15px" id='rep_portuguese_logo_id'>
					<img src="{% static 'insa/SNS_logo.svg' %}" style="width:180px; padding-right:10px" id='sns_logo_id'>
					<img src="{% static 'insa/insa_logo.png' %}" style="width:230px; height:80px; padding-right:12px; padding-top:35px" id='insa_logo_id'>
				</div>
			</div>
			
			<div class="col-sm-3 offset-sm-1 insa-sidebar">
				<img src="{% static 'insa/insaflu_logo.png' %}" style="width:250px; padding-right:25px; padding-top:15px; padding-bottom:15px" id='main_logo_id'>
				<div class="sidebar-module">
					<ol class="list-unstyled" style="font-size: 1.3rem;">
						<li><a href="#" target="_blank">Find here a pre-NGS wet-lab protocol</a></li>
						<li><a href="#" target="_blank">Find here how to design a NGS run</a></li>
					</ol>
          		</div>
				<div class="sidebar-module">
					<h4 class="insa-h4">News</h4>
					<div class="insa-post">
		            	<h2 class="insa-post-title-news">INSaFLU 1.0.0 was launched.</h2>
		            	<p class="insa-post-meta">January 15, 2018 by <a href="#" target="_blank">Vitor Borges</a></p>
		            	<p class="insa-news">The first release of INSaFLU was available for the flu community. This platform supplies public health laboratories
							and influenza researchers with an open “one size fits all” bioinformatics framework that potentiates a global WGS-based
							flu surveillance.</p>
		            	<hr>
		            </div>
          		</div>
				<div class="sidebar-module">
					<h4 class="insa-h4">Useful links</h4>
					<ol class="list-unstyled">
					<li><a href="https://www.fludb.org" target="_blank">Influenza Research Database/Fludb</a></li>
					<li><a href="https://nextflu.org" target="_blank">Nextflu</a></li>
					<li><a href="https://www.gisaid.org/" target="_blank">EpiFLU/GISAID</a></li>
					<li><a href="https://www.ncbi.nlm.nih.gov/genomes/FLU/Database/nph-select.cgi?go=database" target="_blank">NCBI Influenza Virus Resource</a></li>
					<li><a href="http://openflu.vital-it.ch/browse.php" target="_blank">OpenFluDB</a></li>
					<li><a href="https://ecdc.europa.eu/en/seasonal-influenza" target="_blank">ECDC</a></li>
					<li><a href="https://www.cdc.gov/flu/index.htm" target="_blank">CDC</a></li>
					<li><a href="https://microreact.org/showcase" target="_blank">Microreact</a></li>
					<li><a href="https://jameshadfield.github.io/phandango/#/" target="_blank">Phandango</a></li>
					<li><a href="http://www.phyloviz.net/" target="_blank">PHYLOViZ</a></li>
					</ol>
				</div>
				
				<div class="sidebar-module sidebar-module-inset">
					<h6>INSaFLU version 1.0.0</h6>
					<p class="insa-authors">Pipeline developed by Vítor Borges and Miguel Pinheiro.</p>
					<p class="insa-authors">Website by Miguel Pinheiro.</p>
				</div>
				<p> </p>
			</div><!-- /.blog-sidebar -->
		</div> <!-- row -->
	</div>
{% endblock %}